sdk/cosmos/azure-cosmos/azure/cosmos/_gone_retry_policy.py

```diff
@@ -0,0 +1,55 @@
+# The MIT License (MIT)
+# Copyright (c) 2021 Microsoft Corporation
+
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+"""Internal class for connection reset retry policy implementation in the Azure
+Cosmos database service.
+"""
+from . import http_constants
+
+# pylint: disable=protected-access
+
+
+class GoneRetryPolicy(object):
+
+    def __init__(self, client, *args):
+        self._max_retry_attempt_count = 3
+        self.current_retry_attempt_count = 0
+        self.retry_after_in_milliseconds = 1000
+        self.args = args
+        self.client = client
+
+
+    def ShouldRetry(self, exception):
+        """Returns true if should retry based on the passed-in exception.
+
+        :param (exceptions.CosmosHttpResponseError instance) exception:
+        :rtype: boolean
+
+        """
+        if exception.sub_status == http_constants.SubStatusCodes.PARTITION_KEY_RANGE_GONE:
+            # refresh routing_map_provider to refresh partition key range cache
+            # return False to raise error to multi_execution_aggregator and repair document producer
+            self.client.refresh_routing_map_provider()
+            return False
+        if self.current_retry_attempt_count < self._max_retry_attempt_count:
```

Comment:
why are we only checking the current retry attempt count in this if condition, and not when updating the partition key range cache. 
Let's say if we keep getting gone exception, this if condition will never hit, and we will never get out of this retry policy.

Reply:
I was returning a False immediately for sub status 1002 in order to further handle the exception on a first-try basis, since we need the exception to bubble up in order to repair the execution context.

I think this is the way to go, but I see what you mean. I'll add a flag that only returns False pre-emptively if it's the first time we run into a 1002 with that instance of the gone retry policy.